BugFix: improve handling of non-map and too new files in profile's map dir

SlySven opened the pull request:
This is an **alternative** to #5022 - and *one* of them will be needed so that #5016 is safe - though, TBH I do not think that the latter should be dumping JSON format files into a folder which we had previously reserved for binary map save files!

This will also close #4275 by checking for the map version number (the first four bytes of the file) being between 1 and 127 and rejecting it if that test is not met.

Signed-off-by: Stephen Lyons <slysven@virginmedia.com>

#### Release post highlight

Mudlet will behave better if non-map files get placed into the location where a profile saves its map files.

SlySven committed:
BugFix: improve handling of non-map and too new files in profile's map dir

This is an **alternative** to #5022 - and *one* of them will be needed so
that #5016 is safe - though, TBH I do not think that the latter should be
dumping JSON format files into a folder which we had previously reserved
for binary map save files!

This will also close #4275 by checking for the map version number (the
first four bytes of the file) being between 1 and 127 and rejecting it if
that test is not met.

Signed-off-by: Stephen Lyons <slysven@virginmedia.com>

SlySven commented:
Note that a large part of the diff (most of the lower parts) in `TMap.cpp` are due to that part of the code changing its indent as it needed an extra `if (...) {` ... `}`.

In use I have doctored the three most recent files to have version numbers corresponding to `-20`, `22` and `21`, and also inserted a `.png` file with a time-stamp that makes it newer than the more recent *usable* file; the latter is not even considered and the former three are all rejected with an explanation - now an `[ ALERT ] -` rather than an `[ ERROR ] -`:
 
![image](https://user-images.githubusercontent.com/6163092/112875856-a287eb00-90bc-11eb-9911-2f532cd69e8e.png)

SlySven committed:
Revise: keep GHA Clang-tidy appeased

Initialise some variables that are immediately assigned a value in the
next lines following them!

Signed-off-by: Stephen Lyons <slysven@virginmedia.com>

vadi2 commented:
> This is an alternative to #5022 

:+1: what exactly does this do - the description is missing?

SlySven commented:
It means that if `TMap::restore(...)` is called with an *empty* pathFileName it *only* considers files in the profile's `map` subdirectory that end with a (case-insensitive) `.dat` extension and then it iterates through them, starting at the newest file:

It tries to open the file and reads the first four bytes. It checks for a reasonable value (one that maps to an `int32_t` with a value of between 1 and 127) - if the value is:
* outside of that it is rejected with a message given the full file name and saying that it doesn't look to be a Mudlet map file.
* inside that range but it is greater than the current maximum value (`>20`) it is rejected with the message that it is too new, similar to what is done now but without the the check for it being at least feasible (i.e. less than 127) which is what was the big issue in #4275.
* otherwise it is read as before and the iteration stops on that file.

Both of the first actions are as was suggested as a solution to #4275.

In the event that there is no suitable file found the user gets the existing suggestion to either download a map or start one themselves as before.

As it is this will prevent #5016 from causing problems by ignoring `.json` files in that folder - but as I say I am not sure we should be putting the JSON files there anyhow - IMO they should just be placed in the profile's base directory so that the end-user can find them simply and do whatever they want (copy/share/upload?) with them. At the same time I recognise that I do need to address #4808 to complete the JSON map file UI.

vadi2 commented:
Hm, okay. I don't think it's an intuitive design to load .dat map files first over .json ones if they're present in that folder... I like https://github.com/Mudlet/Mudlet/pull/5022 better, so lets get some other opinions in.

vadi2 commented:
As we covered in Discord, let's not make JSON a second-class citizen - saving it in `maps/` is okay and if the latest one is JSON, it should be loaded.

SlySven committed:
Revise: include JSON maps files as ones to try and load

Had to add something so that the (error) texts returned by
`TMap::readJsonMapFile(...)` are translated when used as part of error
messages on the main console.

It does mean that the user could abort the import of a (large) JSON map
file and then it will just try and load the next oldest one. This might
need further work if this proves objectionable.

Signed-off-by: Stephen Lyons <slysven@virginmedia.com>

SlySven committed:
Revise: prevent a JSON map file read on profile loading from being aborted

This is so that JSON map files are treated the same as binary ones in that
situation.

Signed-off-by: Stephen Lyons <slysven@virginmedia.com>

SlySven commented:
Well it works now for JSON files as well as binary ones, apart from ***one*** (🙏 ) glitch - a JSON map gets audited (checked for consistency) **twice** and as that is not a fast action it needs further attention.

SlySven committed:
Revise: prevent double auditing of map when loaded from JSON file

Signed-off-by: Stephen Lyons <slysven@virginmedia.com>

SlySven committed:
Revise: remove iteration though potential map files

Revert to only considering the *newest* file that *could* be a map file.

Signed-off-by: Stephen Lyons <slysven@virginmedia.com>

SlySven committed:
Revise: renumber some QString::arg(...) numbers

Doesn't affect the operation but helps to keep it cleaner.

Signed-off-by: Stephen Lyons <slysven@virginmedia.com>

vadi2 reviewed (approved):
Works great!

Could you make the JSON export be cancellable? It's a pity the binary one isn't also, but it's not something we can address right now.

SlySven commented:
> Could you make the JSON export be cancellable?

I really don't like that as it stops the UI being consistent. I'll do it but grudgingly...

:stuck_out_tongue_closed_eyes:

SlySven committed:
Revise: restore the ability to abort JSON map loading on profile load

Personally I'd have preferred to keep the UI more uniform for both binary
and JSON map file loading but it seems from the peer-review that it was
preferred to be able to abort the load. This will, I think, trigger some
no-map code if the user does abort the loading of a JSON file...

Signed-off-by: Stephen Lyons <slysven@virginmedia.com>

vadi2 commented:
Thank you, appreciate it.
